overlord/snapstate/backend/snapdata.go

```diff
@@ -103,19 +113,70 @@ func (b Backend) untrashData(snap *snap.Info, opts *dirs.SnapDirOptions) error {
 	return err
 }
 
+// removeDirs tries to remove all directories even if there is an error.
+//
+// Note: If there is an error only the first one is returned.
 func removeDirs(dirs []string) error {
+	var firstErr error
 	for _, dir := range dirs {
-		if err := os.RemoveAll(dir); err != nil {
-			return err
+		if err := os.RemoveAll(dir); err != nil && firstErr == nil {
+			firstErr = err
 		}
 	}
 
-	return nil
+	return firstErr
+}
+
+// baseSystemDataHomeDirs returns the per-user base directory for snap data under system.homedirs.
```

Comment:
(nitpick) maybe `configuredSystemDataHomeDirs` or `extraSystemDataHomeDirs` or something? Also `name` -> `snapName` (or `instanceName`) maybe to make it more clear what the context of name is?